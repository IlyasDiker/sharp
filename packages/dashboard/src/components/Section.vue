<script setup lang="ts">
    import { __ } from "@/util/i18n";
</script>

<template>
    <div class="section">
        <div class="row align-items-center">
            <div class="col">
                <SectionTitle
                    :section="section"
                />
            </div>
            <template v-if="hasCommands || filters.length">
                <div class="col-auto align-self-end mb-2" :class="{ 'w-100': filters.length }">
                    <div class="row justify-content-end">
                        <template v-if="filters.length">
                            <div class="col">
                                <div class="row row-cols-auto gx-2">
                                    <template v-for="filter in filters" :key="filter.id">
                                        <SharpFilter
                                            :filter="filter"
                                            :value="filterValue(filter.key)"
                                            @input="$emit('filter-change', filter, $event)"
                                        />
                                    </template>
                                    <template v-if="showReset">
                                        <div class="d-flex">
                                            <button class="btn btn-sm d-inline-flex align-items-center btn-link" @click="$emit('filters-reset', filters)">
                                                {{ __('sharp::filters.reset_all') }}
                                            </button>
                                        </div>
                                    </template>
                                </div>
                            </div>
                        </template>
                        <template v-if="hasCommands">
                            <div class="col-auto">
                                <CommandsDropdown :commands="commands" @select="$emit('command', $event)">
                                    <template v-slot:text>
                                        {{ __('sharp::dashboard.commands.dashboard.label') }}
                                    </template>
                                </CommandsDropdown>
                            </div>
                        </template>
                    </div>
                </div>
            </template>
        </div>

        <Grid :rows="section.rows" row-class="g-3" v-slot="{ itemLayout }">
            <slot :widget-layout="itemLayout" />
        </Grid>
    </div>
</template>

<script lang="ts">
    import {Grid, SectionTitle} from "@sharp/ui";
    import {SharpFilter} from "@sharp/filters";
    import {mapGetters} from "vuex";
    import {CommandsDropdown} from "@sharp/commands";
    import Widget from "./Widget.vue";

    export default {
        components: {
            CommandsDropdown,
            Widget,
            Grid,
            SectionTitle,
            SharpFilter,
        },
        props: {
            section: Object,
            commands: Array,
            filters: Array,
            showReset: Boolean,
        },
        computed: {
            ...mapGetters('dashboard', {
                filterValue: 'filters/value',
            }),
            hasCommands() {
                return this.commands?.flat().length;
            },
        },
    }
</script>
