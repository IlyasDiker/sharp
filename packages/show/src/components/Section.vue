<template>
    <div class="ShowSection" :class="classes">
        <div class="row">
            <template v-if="hasCollapse || section.title">
                <div class="col">
                    <SectionTitle :section="section" :collapsable="hasCollapse" :collapsed="collapsed" />
                </div>
            </template>
            <template v-if="hasCommands && !collapsed">
                <div class="col-auto align-self-end mb-2">
                    <CommandsDropdown :commands="commands" @select="handleCommandSelected">
                        <template v-slot:text>
                            {{ lang('entity_list.commands.instance.label') }}
                        </template>
                    </CommandsDropdown>
                </div>
            </template>
        </div>

        <template v-if="!collapsed">
            <div class="ShowSection__content">
                <Grid class="ShowSection__grid"
                    :rows="[section.columns]"
                    :col-class="() => 'ShowSection__col'"
                    v-slot="{ itemLayout:column }"
                >
                    <Grid class="ShowPage__fields-grid"
                        :rows="column.fields"
                        :row-class="fieldsRowClass"
                        v-slot="{ itemLayout:fieldLayout }"
                    >
                        <slot :field-layout="fieldLayout" />
                    </Grid>
                </Grid>
            </div>
        </template>
    </div>
</template>

<script>
    import { Grid, SectionTitle } from "sharp-ui";
    import { CommandsDropdown } from 'sharp-commands';
    import { lang } from "sharp";

    export default {
        components: {
            Grid,
            CommandsDropdown,
            SectionTitle
        },
        props: {
            section: Object,
            fieldsRowClass: Function,
            collapsable: Boolean,
            layout: {
                type: String,
                validator: layout => ['card', 'contents'].includes(layout),
                default: 'card',
            },
            commands: Array,
        },
        data() {
            return {
                collapsed: this.collapsable,
            }
        },
        computed: {
            classes() {
                return [
                    `ShowSection--layout-${this.layout}`,
                ]
            },
            hasCollapse() {
                return this.collapsable;
            },
            hasCommands() {
                return this.commands?.flat().length > 0;
            },
            wrapperElement() {
                return this.hasCollapse ? 'details' : 'div';
            },
        },
        methods: {
            lang,
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleDetailsToggle(e) {
                this.collapsed = !e.target.open;
            },
        },
    }
</script>
