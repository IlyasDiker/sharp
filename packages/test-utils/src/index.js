export { default as MockInjections } from './MockInjections.vue';
export { default as MockI18n } from './MockI18n.vue';
export { default as QueryComponent } from './QueryComponent.vue';
export { default as MockTransitions } from './MockTransitions.vue';
export { mockSFC, unmockSFC } from './mockSFC.vue';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';
export { createStub } from './stubs';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}
