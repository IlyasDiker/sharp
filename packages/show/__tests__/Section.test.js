import { shallowMount } from '@vue/test-utils';
import Section from "../src/components/Section.vue";

describe('show section', () => {
    function createWrapper(options) {
        return shallowMount(Section, {
            // language=Vue
            stubs: {
                Grid:
                    `<div class="MOCKED_SharpGrid" :class="rowClass && rowClass()">
                        <slot :item-layout="rows[0][0]" />
                    </div>`,
            },
            ...options,
        })
    }

    test('can mount', () => {
        const wrapper = createWrapper({
            propsData: {
                section: {
                    title: 'Section title',
                    columns: [{
                        fields: [[{ key:'name' }]]
                    }]
                },
                fieldsRowClass: () => 'row-class'
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount: collapsable + contents layout', () => {
        const wrapper = createWrapper({
            propsData: {
                section: {
                    title: 'Section title',
                    collapsable: true,
                    columns: [{
                        fields: [[{ key:'name' }]]
                    }]
                },
                collapsable: true,
                layout: 'contents',
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
});
