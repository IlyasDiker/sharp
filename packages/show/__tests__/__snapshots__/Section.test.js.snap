// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show section can mount 1`] = `
<div class="ShowSection ShowSection--layout-card">
	<div class="row">
		<div class="col">
			<div class="ShowSection__header">
				<h2 class="ShowSection__title">Section title</h2>
			</div>
		</div>
		<!---->
	</div>
	<div class="ShowSection__content">
		<div class="MOCKED_SharpGrid ShowSection__grid">
			<div class="MOCKED_SharpGrid ShowPage__fields-grid row-class"></div>
		</div>
	</div>
</div>
`;

exports[`show section can mount: collapsable + contents layout 1`] = `
<div class="ShowSection ShowSection--layout-contents">
	<div class="row">
		<div class="col">
			<details>
				<summary class="ShowSection__header ShowSection__header--collapsable">
					<h2 class="ShowSection__title d-inline-block mb-0">Section title</h2>
				</summary>
			</details>
		</div>
		<!---->
	</div>
	<!---->
</div>
`;
